src/tactic/itauto.lean

```diff
@@ -0,0 +1,457 @@
+/-
+Copyright (c) 2021 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+-/
+import tactic.hint
+
+/-!
+
+# Intuitionistic tautology (`itauto`) decision procedure
+
+The `itauto` tactic will prove any intuitionistic tautology. It implements the well known
+`G4ip` algorithm: <http://www.cs.cmu.edu/~crary/317-f20/homeworks/g4ip.pdf>
```

Comment:
@bryangingechen, do you want this in references.bib?

Reply:
Yeah, that's a good idea.

Reply:
Maybe we should find a better reference? This algorithm is in all the books, I just linked to a random lecture notes here.

Reply:
Most of the sources for G4ip point to some obscure work by Vorob'ev that I can't get a hold of (and is in Russian I think), "the derivability problem in the constructive propositional calculus with strong negations". Which Dyckhoff paper are you thinking of?

Reply:
The name G4ip is obviously due to a very well-known textbook by Troelstra and Schwichtenberg.  The book is definitely worth a read, and uses this abbreviation scheme extensively: there's also G2i, G3c, ....  This is what they have to say about G4ip:
> 4.9.4. The Gentzen system G4ip. This calculus, with its splitting of L→,
> was recently discovered independently by Dyckhoff [1992] and Hudelmaier
> [1989,1992]. Here we have combined the notion of weight as defined by Hudel-
> maier with Dyckhoff's clever proof of equivalence with the ordinary calculus.
> Long before these recent publications, a decision algorithm based on the same
> or very similar ideas appeared in a paper by Vorob'ev [1964]; but in that pa-
> per the present formalism is not immediately recognizable. Remarks on the
> history of this calculus may be found in Dyckhoff's paper.

The Dyckhoff paper is titled "Contraction-free sequent calculi for intuitionistic logic".